documentcloud/documents.py

```diff
@@ -445,6 +553,21 @@ def upload_directory(self, path, handle_errors=False, extension=".pdf", **kwargs
         # Pass back the list of documents
         return [Document(self.client, d) for d in obj_list]
 
+    def _has_valid_extension(self, filename, extensions, supported_extensions):
+        """Check if the file has a valid extension"""
+        if extensions is None:
+            return True
+
+        # Ensure the provided extensions are from the approved list
```

Comment:
You don't need to check `extensions` for every file - it is constant for the whole `upload_directory` call.  Lift this check up to that function.